resourceIndexer/coreStagingS3.js

```diff
@@ -0,0 +1,294 @@
+import { SOURCE, VALID_AUSPICE_PATTERNS, INVALID_AUSPICE_PATTERNS,
+  DATESTAMP_REGEX, SIDECAR_TYPES } from './constants.js';
+import { collectInventory } from './inventory.js';
+
+/**
+ * The inventory of buckets (especially the core bucket) is in some ways a
+ * historical record of work over the years, but this isn't really what we want
+ * to display to users. As some examples:
+ *    - Files which don't match a resource to list should be excluded
+ *    - Datestampted files (i.e. _YYYY-MM-DD in the filename) are excluded
+ *      (we use S3 versioning instead)
+ * 
+ * If the s3 object is to be excluded we return false here.
+ * 
+ * In the case where the object represents a (part of) a resource we want to
+ * expose, then we categorise it here by adding the following properties:
+ *   - source (STAGING or CORE)
+ *   - resourceType (dataset, narrative or intermediate)
+ *   - id (the ID by which objects will be grouped together.
+ *        For datasets this is the nextstrain.org URL path, without any temporal signifier)
+ *   - subresourceType (currently only for resourceType=dataset)
+ */
+function categoriseCoreObjects(item, staging) {
+  const key = item.key;
+  item.source = staging ? SOURCE.STAGING : SOURCE.CORE;
+  item.baseUrl = `https://${item.bucket}.s3.amazonaws.com/${key}`
+  if (key.startsWith('search_')
+    || key.startsWith('manifest_')
+    || key.startsWith('datasets_')
+  ) return false;
+
+  // On the core bucket, directory-like hierarchies are used for intermediate
+  // files. These intermediate files may include files which auspice can
+  // display, but nextstrain.org cannot map URLs to directory-like hierarchies.
+  // There are other resourceTypes here we may consider in the future -- e.g.
+  // model output JSONs
+  if (key.includes("/")) {
+    if (staging===true) return false;
+    if (key.startsWith('files/')) {
+      if (
+        key.includes('/archive/')
+        || key.includes('/test/')
+        || key.includes('/workflows/')
+        || key.includes('/branch/')
+        || key.includes('/trial/')
+        || key.includes('/test-data/')
+        || key.includes('jen_test/')
+        || key.match(/\/nextclade-full-run-[\d-]+--UTC\//)
+        || key.match(/\/\d{4}-\d{2}-\d{2}_results.json/) // forecasts-ncov
+        || key.endsWith('.png')                          // forecasts-ncov
+      ) {
+        return false;
+      }
+      item.resourceType = 'intermediate';
+      /* The ID is used for grouping. For a nextstrain.org dataset this would be
+      combined with the source to form a nextstrain URL, however that's not
+      applicable here. Instead we use the filepath information without the
+      leading 'files/' and without the (trailing) filename so that different
+      files in the same directory structure get grouped together. For instance,
+      files/ncov/open/x.json -> ncov/open */
+      const fields = key.split('/')
+      item.resourcePath = fields.slice(1, fields.length-1).join('/')
```

Comment:
In the context of a PR with little commentary, a suggestion like this is useful: no functional changes, but 2 lines down to 1. In the context of a PR with hundreds of comments this suggestion (and others) seem overly nitpicky, and from my point of view hinder the momentum of the work.

In order to facilitate the progress of the PR I've switched to your suggestion.

Reply:
I'm sorry if your momentum was hindered by this suggestion or others. It was a small suggestion intended for you to take or leave as you saw fit. Dismissing it would not have impeded the progress of the PR from my point of view.

I see a large part of code review as akin to the job of a editor (as in publishing), and anything that tightens up the writing—makes it clearer, less wordy, more idiomatic, etc—as worth suggesting.